elf_reader.go

```diff
@@ -523,14 +523,16 @@ func uintFromBTF(typ btf.Type) (uint32, error) {
 }
 
 func (ec *elfCode) loadDataSections(maps map[string]*MapSpec, dataSections map[elf.SectionIndex]*elf.Section, spec *btf.Spec) error {
-	if spec == nil {
-		return xerrors.New("data sections require BTF")
```

Comment:
Okay, I figured out what about my program causes LLVM to emit an `.rodata` section. My program contains `const int ready_val = 0x90;`, as someone would naively write in C.

```
~ readelf -a build/bpf/acct/5.5.10.o
...
Section Headers:
  [Nr] Name              Type             Address           Offset
       Size              EntSize          Flags  Link  Info  Align
...
  [11] .rodata           PROGBITS         0000000000000000  00001890
       0000000000000004  0000000000000000   A       0     0     4
```

Dumping the contents with `sec.Data()` in `loadDataSections()`, .rodata contains: `[144 0 0 0]`, or hex 0x90.

I'm not quite sure why LLVM needs to put this into a data section, because it's only ever used in comparisons like follows:
```
11:       55 01 14 01 90 00 00 00 if r1 != 144 goto +276 <LBB0_51>
```

This might be something that needs to be addressed in the BPF program itself, but without the nil check on `*btf.Spec` in `loadDataSections`, a call to `spec.Datasec()` will obviously panic. I'm not sure what the right path forward is here. We could just silently skip loading data sections if there's no BTF spec available, but that might lead to hard-to-debug surprises down the road.

Reply:
The compiler can't eliminate `ready_val` because it's not marked as `static` :(

I guess you didn't want a data section, but got one due to missing `static` and thought the error to load is due to missing BTF?

Reply:
Yup! I'd like to extend the error message here to at least include the section name: `found data section %s, but the ELF binary did not contain a required BTF spec`.

Could we maybe add an extra message in case the offending section is `.rodata` to hint at consts with missing `static`? I have a feeling this might become a common issue for users. :)